plugin/transfer/transfer.go

```diff
@@ -0,0 +1,135 @@
+package transfer
+
+import (
+	"context"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+var log = clog.NewWithPlugin(pluginName)
+
+// Transfer is a plugin that handles zone transfers.
+type Transfer struct {
+	xfrs []*xfr
+	Next plugin.Handler
+}
+
+type xfr struct {
+	Zones       []string
+	to          []string
+	Transferers map[string]Transferer // a map of zone->plugins which implement Transferer
+}
+
+// Transferer may be implemented by plugins to enable zone transfers
+type Transferer interface {
+	// AllRecords writes all records in the zone to the channel.  The
+	// first and last records written to the channel must be the SOA for the zone.
+	// If serial is non zero, an incremental zone transfer should be assumed.
+	AllRecords(zone string, serial uint32, ch chan dns.RR)
```

Comment:
I don't like this interface yet. I think returning a channel is more natural, also the channel should be recieve only, not the send/receive it is now.
I.e. 
~~~
Transfer(zone string, serial uint32) (ch <-chan []dns.RR)
~~~
This also means the plugin closes the channel after is sent the last rrrs, which is more natural, not sure how that would work now.
And allowing a slice is also way more efficient, because that's probably what most plugins are holding anyway. Also sending the SOA twice seems axfr implemention detail and not super relevant to the implementation. The transfer plugin can take care of that.